File: AQUÍ/PerceptronJava/Perceptron.java
import java.util.function.Function;

public class Perceptron {
    private double[] pesos;
    private double bias;
    private double tasaAprendizaje;
    private Function<Double, Double> activacion;
    private String nombreActivacion;

    // Constructor
    public Perceptron(int numEntradas, double tasaAprendizaje, String tipoActivacion) {
        this.pesos = new double[numEntradas];
        this.bias = 0.0;
        this.tasaAprendizaje = tasaAprendizaje;
        
        // Inicializar pesos aleatorios entre -1 y 1
        for (int i = 0; i < numEntradas; i++) {
            this.pesos[i] = Math.random() * 2 - 1;
        }
        
        this.activacion = seleccionarActivacion(tipoActivacion);
        this.nombreActivacion = tipoActivacion;
    }
    
    // Seleccionar función de activación
    private Function<Double, Double> seleccionarActivacion(String tipo) {
        switch (tipo.toLowerCase()) {
            case "lineal":
                return x -> x;
            case "escalon":
                return x -> x >= 0 ? 1.0 : 0.0;
            case "sigmoidal":
                return x -> 1.0 / (1.0 + Math.exp(-x));
            case "relu":
                return x -> Math.max(0, x);
            case "softmax":
                // Para clasificación binaria, softmax se comporta como sigmoide
                return x -> 1.0 / (1.0 + Math.exp(-x));
            case "tanh":
                return x -> Math.tanh(x);
            default:
                return x -> x >= 0 ? 1.0 : 0.0;
        }
    }
    
    // Calcular sumatoria ponderada
    private double calcularSumatoria(double[] entradas) {
        double suma = bias;
        for (int i = 0; i < entradas.length; i++) {
            suma += pesos[i] * entradas[i];
        }
        return suma;
    }
    
    // Predecir salida para una entrada
    public double predecir(double[] entradas) {
        double sumatoria = calcularSumatoria(entradas);
        return activacion.apply(sumatoria);
    }
    
    // Entrenar el perceptrón
    public void entrenar(double[][] datos, double[] etiquetas, int iteraciones) {
        System.out.println("Entrenando con " + nombreActivacion + "...");
        
        for (int iter = 0; iter < iteraciones; iter++) {
            double errorTotal = 0;
            
            for (int i = 0; i < datos.length; i++) {
                double prediccion = predecir(datos[i]);
                double error = etiquetas[i] - prediccion;
                errorTotal += Math.abs(error);
                
                // Actualizar pesos y bias
                for (int j = 0; j < pesos.length; j++) {
                    pesos[j] += tasaAprendizaje * error * datos[i][j];
                }
                bias += tasaAprendizaje * error;
            }
            
            // Mostrar progreso cada 10 iteraciones
            if (iter % 10 == 0 || iter == iteraciones - 1) {
                System.out.printf("Iteración %d: Error total = %.4f%n", iter, errorTotal);
            }
        }
        System.out.println("Entrenamiento completado\n");
    }
    
    // Evaluar precisión
    public double evaluar(double[][] datos, double[] etiquetas) {
        int aciertos = 0;
        for (int i = 0; i < datos.length; i++) {
            double prediccion = predecir(datos[i]);
            // Redondear para comparar con etiquetas (por si es sigmoide)
            int clasePredicha = (prediccion >= 0.5) ? 1 : 0;
            if (clasePredicha == (int) etiquetas[i]) {
                aciertos++;
            }
        }
        return (double) aciertos / datos.length * 100;
    }
    
    // Mostrar pesos finales
    public void mostrarPesos() {
        System.out.print("Pesos finales: [");
        for (int i = 0; i < pesos.length; i++) {
            System.out.printf("%.4f", pesos[i]);
            if (i < pesos.length - 1) System.out.print(", ");
        }
        System.out.printf("], Bias: %.4f%n", bias);
    }
}

File: PerceptronJS/perceptron.js
class Perceptron {
    constructor(numEntradas, tasaAprendizaje, tipoActivacion) {
        this.pesos = [];
        this.bias = 0;
        this.tasaAprendizaje = tasaAprendizaje;
        this.tipoActivacion = tipoActivacion;
        
        // Inicializar pesos aleatorios entre -1 y 1
        for (let i = 0; i < numEntradas; i++) {
            this.pesos.push(Math.random() * 2 - 1);
        }
        
        // Asignar función de activación
        this.activacion = this.seleccionarActivacion(tipoActivacion);
    }
    
    // Seleccionar función de activación
    seleccionarActivacion(tipo) {
        switch (tipo.toLowerCase()) {
            case 'lineal':
                return (x) => x;
            case 'escalon':
                return (x) => x >= 0 ? 1 : 0;
            case 'sigmoidal':
                return (x) => 1 / (1 + Math.exp(-x));
            case 'relu':
                return (x) => Math.max(0, x);
            case 'softmax':
                // Para clasificación binaria, softmax se comporta como sigmoide
                return (x) => 1 / (1 + Math.exp(-x));
            case 'tanh':
                return (x) => Math.tanh(x);
            default:
                return (x) => x >= 0 ? 1 : 0;
        }
    }
    
    // Calcular sumatoria ponderada
    calcularSumatoria(entradas) {
        let suma = this.bias;
        for (let i = 0; i < entradas.length; i++) {
            suma += this.pesos[i] * entradas[i];
        }
        return suma;
    }
    
    // Predecir salida para una entrada
    predecir(entradas) {
        const sumatoria = this.calcularSumatoria(entradas);
        return this.activacion(sumatoria);
    }
    
    // Entrenar el perceptrón
    entrenar(datos, etiquetas, iteraciones) {
        console.log(`Entrenando con ${this.tipoActivacion}...`);
        
        for (let iter = 0; iter < iteraciones; iter++) {
            let errorTotal = 0;
            
            for (let i = 0; i < datos.length; i++) {
                const prediccion = this.predecir(datos[i]);
                const error = etiquetas[i] - prediccion;
                errorTotal += Math.abs(error);
                
                // Actualizar pesos y bias
                for (let j = 0; j < this.pesos.length; j++) {
                    this.pesos[j] += this.tasaAprendizaje * error * datos[i][j];
                }
                this.bias += this.tasaAprendizaje * error;
            }
            
            // Mostrar progreso cada 10 iteraciones
            if (iter % 10 === 0 || iter === iteraciones - 1) {
                console.log(`Iteración ${iter}: Error total = ${errorTotal.toFixed(4)}`);
            }
        }
        console.log('Entrenamiento completado\n');
    }
    
    // Evaluar precisión
    evaluar(datos, etiquetas) {
        let aciertos = 0;
        for (let i = 0; i < datos.length; i++) {
            const prediccion = this.predecir(datos[i]);
            // Redondear para comparar con etiquetas (por si es sigmoide)
            const clasePredicha = prediccion >= 0.5 ? 1 : 0;
            if (clasePredicha === etiquetas[i]) {
                aciertos++;
            }
        }
        return (aciertos / datos.length) * 100;
    }
    
    // Mostrar pesos finales
    mostrarPesos() {
        const pesosStr = this.pesos.map(p => p.toFixed(4)).join(', ');
        console.log(`Pesos finales: [${pesosStr}], Bias: ${this.bias.toFixed(4)}`);
    }
}

// Exportar para Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = Perceptron;
}